### DISCUSSION TITLE
./configure --without-bonmin not being honored: configure: error: Required package Bonmin not available

### DISCUSSION THREAD

**Discussion by LinuxPersonEC** [coin-or/Couenne#74] [General]
On RHEL 8 is this a bug or intentional? 

```
./configure   --without-bonmin --prefix=/path/to/me/.local
checking how to hardcode library paths into programs... immediate
checking for pkgconf... pkgconf
checking pkgconf is at least version 0.16.0... yes: 1.4.2
configure: pkgconf path is "/path/to/me/.local/lib/pkgconfig:"
checking for package Bonmin... skipping
configure: error: Required package Bonmin not available.
```

And a different issue since Cgl is required that fails to make with:
```
CglGomory.cpp:33:10: fatal error: OsiClpSolverInterface.hpp: No such file or directory
 #include "OsiClpSolverInterface.hpp"
```

**Comment by merraksh** [comment]
On any architecture/platform, Couenne needs Bonmin as it builds on its branch-and-bound scheme.

**Reply by LinuxPersonEC** [reply]
Then why is that option available?

I can't get Bonmin to install since Cgl fails with the aforementioned error. Any ideas?

**Reply by merraksh** [reply]
I think that option is generated by default for most dependencies, but it should be dropped. I'm afraid I have no idea how to fix the Bonmin/Cgl issue, maybe try the Bonmin discussion page.

**Reply by LinuxPersonEC** [reply]
On https://www.coin-or.org/Bonmin/ the mailing list link goes to a broken link https://www.coin-or.org/mail.html

I posted in the Git page as well as Cgl Git, no responses yet.

**Reply by LinuxPersonEC** [reply]
I found a [Docker container](https://github.com/TaskeHAMANO/PyomoContainer/blob/master/Dockerfile) that can be pulled via Singularity. 
It includes:

- Bonmin
- CBC
- Couenne
- GLPK
- IpOpt
- SCIP 

The container is in the [Docker repository](https://hub.docker.com/r/koshoshiyuji/pyomocontainer_jupyter) last updated July 2020.

So in order to use it, I loaded Singularity:
```
module load singularity
singularity pull docker://koshoshiyuji/pyomocontainer_jupyter:mod21
```
This creates a .sif (Singularity Image Format) container called `pyomocontainer_jupyter_mod21.sif.` Note that the container .sif file is ~ 1.8 GB so it takes a few minutes to pull.

I downloaded [2 .nl files to test](https://github.com/cvanaret/Uno/tree/main/examples) with Couenne

Then I ran `couenne` against those 2 `.nl` files:

`singularity shell pyomocontainer_jupyter_mod21.sif`

Which I followed up with:
```
Singularity> couenne [polak5.nl](http://polak5.nl/) 
Couenne 0.5.7 -- an Open-Source solver for Mixed Integer Nonlinear Optimization
Mailing list: [couenne@list.coin-or.org](mailto:couenne@list.coin-or.org)
Instructions: http://www.coin-or.org/Couenne

NLP0012I 
Num Status Obj It time Location
NLP0014I 1 OPT 50 6 0.001931
Couenne: new cutoff value 4.9999999990e+01 (0.006037 seconds)
Loaded instance "[polak5.nl](http://polak5.nl/)"
Constraints: 2
Variables: 3 (0 integer)
Auxiliaries: 8 (0 integer)

Coin0506I Presolve 21 (-1) rows, 9 (-2) columns and 46 (-3) elements
Clp0006I 0 Obj 0 Primal inf 9.0431615 (7)
Clp0006I 11 Obj 50
Clp0000I Optimal - objective value 50
Clp0032I Optimal objective 49.99999987 - 11 iterations time 0.002, Presolve 0.00
Clp0000I Optimal - objective value 50
Cbc0012I Integer solution of 50 found by Couenne Rounding NLP after 0 iterations and 0 nodes (0.00 seconds)
NLP Heuristic: NLP0014I 2 OPT 50 41 0.006194
no solution.
Cbc0001I Search completed - best objective 49.99999999001755, took 0 iterations and 0 nodes (0.01 seconds)
Cbc0035I Maximum depth 0, 0 variables fixed on reduced cost
Clp0006I 0 Obj 50 Primal inf 0.54184964 (1)
Clp0006I 1 Obj 50
Clp0000I Optimal - objective value 50
Couenne: new cutoff value 4.9999999870e+01 (0.014454 seconds)

couenne: Optimal

"Finished"

Linearization cuts added at root node: 22
Linearization cuts added in total: 22 (separation time: 5e-06s)
Total solve time: 0.008027s (0.008027s in branch-and-bound)
Lower bound: 50
Upper bound: 50 (gap: 0.00%)
Branch-and-bound nodes: 0
Performance of FBBT: 2e-06s, 3 runs. fix: 0 shrnk: 0 ubd: 4.33333 2ubd: 2.66667 infeas: 0
```
Then another:

```
Singularity> couenne [hs015.nl](http://hs015.nl/) 
Couenne 0.5.7 -- an Open-Source solver for Mixed Integer Nonlinear Optimization
Mailing list: [couenne@list.coin-or.org](mailto:couenne@list.coin-or.org)
Instructions: http://www.coin-or.org/Couenne

NLP0012I 
Num Status Obj It time Location
NLP0014I 1 OPT 306.49998 22 0.005468
Couenne: new cutoff value 3.0649997900e+02 (0.009954 seconds)
Loaded instance "[hs015.nl](http://hs015.nl/)"
Constraints: 2
Variables: 2 (0 integer)
Auxiliaries: 8 (0 integer)

Coin0506I Presolve 29 (-1) rows, 9 (-1) columns and 64 (-2) elements
Clp0006I 0 Obj 0.25 Primal inf 473.75936 (14)
Clp0006I 13 Obj 0.31728151
Clp0000I Optimal - objective value 0.31728151
Clp0032I Optimal objective 0.3172815065 - 13 iterations time 0.002, Presolve 0.00
Clp0000I Optimal - objective value 0.31728151
Cbc0012I Integer solution of 306.49998 found by Couenne Rounding NLP after 0 iterations and 0 nodes (0.00 seconds)
NLP Heuristic: NLP0014I 2 OPT 306.49998 5 0.001204
solution found, obj. 306.5
Clp0000I Optimal - objective value 0.31728151
Optimality Based BT: 3 improved bounds
Probing: 2 improved bounds
Cbc0031I 1 added rows had average density of 2
Cbc0013I At root node, 4 cuts changed objective from 0.31728151 to 306.49998 in 1 passes
Cbc0014I Cut generator 0 (Couenne convexifier cuts) - 4 row cuts average 2.0 elements, 3 column cuts (3 active)
Cbc0001I Search completed - best objective 306.4999790004336, took 0 iterations and 0 nodes (0.00 seconds)
Cbc0035I Maximum depth 0, 0 variables fixed on reduced cost

couenne: Optimal

"Finished"

Linearization cuts added at root node: 30
Linearization cuts added in total: 30 (separation time: 0.000106s)
Total solve time: 0.004338s (0.004338s in branch-and-bound)
Lower bound: 306.5
Upper bound: 306.5 (gap: 0.00%)
Branch-and-bound nodes: 0
Performance of FBBT: 6.9e-05s, 4 runs. fix: 0 shrnk: 0.000103838 ubd: 2.75 2ubd: 0.5 infeas: 0
Performance of OBBT: 7e-05s, 1 runs. fix: 0 shrnk: 6.70203 ubd: 0 2ubd: 0 infeas: 0
```

Any other suggestions for tests/tutorials I can try?

**Reply by tkralphs** [reply]
I apologize for the late reply, but I wanted to respond to a couple if different things mentioned here. 

- First, yes, we have m4 macros that automate a lot of what one needs to do to deal with dependencies and those automatically generate the `--without-*` options. I don't think there is a way to specify that this should not be done, but it is a good idea. Surprisingly, it has never been suggested before now. There is actually a new version of our build tools, which is currently deployed in Couenne's master branch, but I think the behavior is he same there. 
- For the Cgl issue, I have now seen it and will respond there (you will notice that the same people tend to pop up in discussions for many of the projects---there is a very small group of people maintaining many of the legacy projects),
- For the Bonmin mailing list, the right place to post questions now is Github Discussions. One of these days, I will get around to modifying the README to reflect this. 
- For Docker, there is also a container that I maintain here: https://hub.docker.com/repository/docker/coinor/coin-or-optimization-suite. I'm not sure there's any advantage to it over what you already found, but it is the "official" container for Couenne. This is another thing that should be in the README that will be added "one of these days."
- For tests/tutorials, what are you looking for exactly? Looks like everything is working well.

**Comment by LinuxPersonEC** [comment]
> * For tests/tutorials, what are you looking for exactly? Looks like everything is working well.

Well since `WORKDIR` is set to `root`, I don't think it's possible to open the JupyterLab will ever open, as we use Singularity, so the real ask is a Singularity-friendly container.

I'll test `singularity pull  docker://coinor/coin-or-optimization-suite` now.

Thanks for the reply and link to the official container.

As I mentioned, the web site also has the broken link to the mailing list as does the afore-mentioned container:

```
Singularity> couenne [hs015.nl](http://hs015.nl/) 
Couenne 0.5.7 -- an Open-Source solver for Mixed Integer Nonlinear Optimization
Mailing list: [couenne@list.coin-or.org](mailto:couenne@list.coin-or.org)
Instructions: http://www.coin-or.org/Couenne
```

I reached out to the person who created the "unofficial" container to ask as well, at least about the root user.

**Reply by tkralphs** [reply]
>As I mentioned, the web site also has the broken link to the mailing list as does the afore-mentioned container:

Yes, this will all be fixed "soon." I maintain a lot of projects and it's whack-a-mole keeping stuff like this up to date. I have scripts for generating up-to-date READMEs for many of the projects, but haven't folded Couenne into that workflow yet. It's a bit of work.
